cmd/webhook/webhook.go

```diff
@@ -135,6 +135,17 @@ func patchCPULimit(sharedCPUTime int, patchList []patch, i int, c *corev1.Contai
 		patchItem.Value = json.RawMessage(`{ "limits": { "cpu":` + cpuVal + ` } }`)
 	}
 	patchList = append(patchList, patchItem)
+	patchItem.Op = "add"
+
+	cpuVal = `"0m"`
+	if len(c.Resources.Limits) > 0 {
```

Comment:
Here we should check for `c.Resources.Requests`. Above we are checking for `Limits` but that shouldn't be needed as pool resource request must be specified in pod manifest in order to get cpu(s) from the pool. So the `resources/limits` should be there always.  And it seems that when cpu `limits` is given in manifest, the similar `requests` is added to pod specification (by Kubernetes?).  So based on this we should have `limits` and `requests` always in the pod specification when we start to create patches. 

As a conclusion both of those checks could be removed.

Reply:
see https://github.com/nokia/CPU-Pooler/issues/14 for why we actually need to make sure requests need to be zeroed out for shared pool
yes, for the Device requests / limits should be there, but this is (or should,  I haven't checked so far the review :) ) related to the core CPU request added at the same time
if we allow a container to define core CPU request while also asking for shred devices, we would actually lower the same amount also from the default pool. that would be a pure waste

Reply:
after checking the review: due to above I think the check is still needed, but it should be based on the exact condition, not on the effect on the condition
i.e. we need to add cpu.limit and zero out cpu.request when the container is put onto a shared pool, instead of assuming "we have added limit **because** it is a shared pool container"

Reply:
There are two different patches created for `requests.limits`. One for the case when `limits` exists already and one for the case when `limits` does not exist. The patch is different in these cases. Please note that it's not about `cpu` limits, it can be something else. In our case when pod asks cpus from a pool, it must define the pool request in the `resources` structure. So we know the `limits` exists in the pod manifest and we know what kind of patch we can create. The patch is created anway for `cpus`

Reply:
I feel we are not on the same page still :)
Let's go one by one, and see where we differ.

> Please note that it's not about cpu limits, 
This is review is just about the CPU limits, and requests. And only when resources were requested from the shared CPU pool.
Patch CPU limit is only invoked in case shared resources are requested:
https://github.com/nokia/CPU-Pooler/blob/b222d26d8f66ff1140dbe8a2c50857ad1a978bc2/cmd/webhook/webhook.go#L253

> The patch is created anway for cpus
We create the patch for CPUs within this function, and the review is about that we shall do it differently.
The current code actually provisions CPU limit whether there was a request or not, because in both branches /resources/limits/**cpu** is getting patched (lines 130 and 135, just in different ways).
This is okay functionally, but the logic could be simplified to simply patch unconditionally.
Bálint just tried to copy the same logic when patching /resources/requests/**cpu** - which is again overflous, we should just simply unconditionally patch 0 into the request/cpu.

But in any case: we are the one patching default CPU related limits and requests, and we are only patching CPU requests and limits here. And even that we should do in simpler way

Reply:
ok, I think I now see what you mean, but please note that the patch type is "add" in both cases, not "replace". I'm not sure even in its current form the patch would really overwrite the whole request/limit structure

in any case, we really only need two simply, unconditional patches IMO in this function: replace requests/cpu with 0, and replace limits/cpu with the requested amount of shared CPU
we can probably leave rest of the requests/limits untouched in all cases

Reply:
(I think a replace "cpu/request" would work even if "request" is not added by default, but we can test it out)

Reply:
My understanding is that replace does not work if `cpu` member exists already. And according to RFC6902 the `add` should replace the existing member if it already exists. But it's good if you can verify it with a test.

Reply:
yeah Bálint is doing that in the meantime.
I just wrote a hook for DANM, and I used "replace" for all patches. It looks to be implicitly doing an add when it is required, otherwise it replaces the whole dictionary.
I actually had a problem with the opposite: when I just added a key to an existing dictionary my key got duplicated. Maybe it was just because I was mutating a CRD, but generally I found using replace safer